drivers/clock_control/clock_control_renesas_ra_cgc.c

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2024 Renesas Electronics Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/kernel.h>
+#include <soc.h>
+#include <zephyr/dt-bindings/clock/ra_clock.h>
+#include <zephyr/drivers/clock_control/renesas_ra_cgc.h>
+
+static int clock_control_renesas_ra_on(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct clock_control_ra_subsys_cfg *subsys_clk = (struct clock_control_ra_subsys_cfg *)sys;
+
+	if (!dev || !sys) {
+		return -EINVAL;
+	}
+	WRITE_BIT(*subsys_clk->mstp, subsys_clk->stop_bit, false);
+	return 0;
+}
+
+static int clock_control_renesas_ra_off(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct clock_control_ra_subsys_cfg *subsys_clk = (struct clock_control_ra_subsys_cfg *)sys;
+
+	if (!dev || !sys) {
+		return -EINVAL;
+	}
+
+	WRITE_BIT(*subsys_clk->mstp, subsys_clk->stop_bit, true);
+	return 0;
+}
+
+static int clock_control_renesas_ra_get_rate(const struct device *dev, clock_control_subsys_t sys,
+					     uint32_t *rate)
+{
+	const struct clock_control_ra_pclk_cfg *config = dev->config;
```

Comment:
https://github.com/zephyrproject-rtos/zephyr/pull/75254#discussion_r1692950049

As mentioned in 75254, the `get_rate` should be implemented appropriately.
This is a general Zephyr method to allow each driver to properly reference the frequency of the clock source specified in the DeviceTree. It should be implemented in the Zephyr way without relying on the FSP mechanism.

Reply:
`In Zephyr, it is expected to call the get_rate of clock_controller with the value of clocks for each node set in the device tree to obtain the clock frequency required by that peripheral. The API of clock_controller is designed with this usage in mind.`

@soburi , I'm aware of this mechanism in Zephyr, as I mention before this implementation is aiming to treat each pclk as a clock device on Zephyr and peripheral node can choose them as clocks and get their information by either clock subsys or clock device config. Please help to check it's usage first in the link below:

![image](https://github.com/user-attachments/assets/d26a45d7-c92d-42d0-8b32-0da5115532dc)
![image](https://github.com/user-attachments/assets/224d0aea-83b4-436e-b11e-b8f7da5f5270)
![image](https://github.com/user-attachments/assets/37dfc4b9-401d-41de-83d8-1ec5e73b6bc8)

Here canclk node will be intialize as a clock device on Zephyr and it's added to CAN clocks cell so that clock DLL clock can be initialize and rate can be get from it. It'll be similar case for almost all other module

https://github.com/zephyrproject-rtos/zephyr/pull/76134/files#diff-5593f239f87a26980936c420f958bd0bfd283dafe6af955ecb412937855e410eR297
https://github.com/zephyrproject-rtos/zephyr/pull/76134/files#diff-a5bf8cca928774ccefddc926c4076c0471caddd611a1e5991fa9e99288419a0eR1117-R1125
https://github.com/zephyrproject-rtos/zephyr/pull/76134/files#diff-a5bf8cca928774ccefddc926c4076c0471caddd611a1e5991fa9e99288419a0eR944

Reply:
Ah, I see.
So there are multiple clock-controllers which has "renesas,ra8-cgc-pclk" compatible property.
This certainly works, but at least it's a pretty original implementation.

@nordic-krch
As far as I know, there are no devices that normally create multiple clock-controllers.
This is a proposal to create an instance of a clock controller for each clock source, but I want to hear your opinion.